R4R:  Sunny/tombstone spec (#3103)
cosmos/cosmos-sdk: docs/spec/slashing/tombstone.md

4. Validator A unjails
   
We are now in a new slashing period, however we still have to keep the door open for the previous infraction, as the evidence for Infraction 2 may still come in. As the number of slashing periods increase, it creates more complexity as we have to keep track of the highest infraction amount for every single slashing period.

> Note:  Currently, according to the `slashing` module spec, a new slashing period is created every time a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed.  See issue [#3205](https://github.com/cosmos/cosmos-sdk/issues/3205) for further details.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.

The maximum number of slashing periods is the `len(UnbondingPeriod) / len(JailPeriod)`.  The current defaults in Gaia for the `UnbondingPeriod` and `JailPeriod` are 3 weeks and 2 days, respectively.  This means there could potentially be up to 11 slashing periods concurrently being tracked per validator.  If we set the `JailPeriod >= UnbondingPeriod`, we only have to track 1 slashing period (i.e not have to track slashing periods).

> Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose setting the "jail time" for a validator who commits a consensus safety fault, to `infinite` (i.e. a tombstone state).  This essentially kicks the validator out of the validator set and does not allow them to re-enter the validator set.  All of their delegators (including the operator themselves) have to either unbond or redelegate away.  The validator operator can create a new validator if they would like, with a new operator key and consensus key, but they have to "re-earn" their delegations back.  To put the validator in the tombstone state, we set `DoubleSignJailEndTime` to `time.Unix(253402300800)`, the MAX time supported by Amino.

Reviewer: ```suggestion
Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose setting the "jail time" for a validator who commits a consensus safety fault, to `infinite` (i.e. a tombstone state).  This essentially kicks the validator out of the validator set and does not allow them to re-enter the validator set.  All of their delegators (including the operator themselves) have to either unbond or redelegate away.  The validator operator can create a new validator if they would like, with a new operator key and consensus key, but they have to "re-earn" their delegations back.  To put the validator in the tombstone state, we set `DoubleSignJailEndTime` to `time.Unix(253402300800)`, the maximum time supported by Amino.
```

Implementing the tombstone system and getting rid of the slashing period tracking will make the `slashing` module way simpler, especially because we can remove all of the hooks defined in the `slashing` module consumed by the `staking` module (the `slashing` module still consumes hooks defined in `staking`).

### Single slashing amount

Another optimization that can be made is that if we assume that all ABCI faults for Tendermint consensus are slashed at the same level, we don't have to keep track of "max slash".  Once an ABCI fault happens, we don't have to worry about comparing potential future ones to find the max.

Currently the only Tendermint ABCI fault is:
- Unjustified precommits (double signs)

> In it currently planned to include the following fault in the near future:

Reviewer: ```suggestion
It is currently planned to include the following fault in the near future:
```

# Staking Tombstone

## Abstract

In the current implementation of the `slashing` module, when the consensus engine informs the state machine of a validator's consensus fault, the validator is partially slashed, and put into a "jail period", a period of time in which they are not allowed to rejoin the validator set.  However, because of the nature of consensus faults and ABCI, there can be a delay between an infraction occuring, and evidence of the infraction reaching the state machine (this is one of the primary reasons for the existence of the unbonding period).

> > Note: The tombstone concept, only applies to faults that have a delay between the infraction occuring and evidence reaching the state machine.  For example, evidence of a validator double signing may take a while to reach the state machine due to unpredictable evidence gossip layer delays and the ability of validators to selectively reveal double-signatures (e.g. to infrequently-online light clients).  Liveness slashing, on the other hand, is detected immediately as soon as the infraction occurs, and therefore no slashing period is needed.  A validator is immediately put into jail period, and they cannot commit another liveness fault until they unjail.  In the future, there may be other types of byznatine faults that have delays (for example, submitting evidence of an invalid proposal as a tranasction).  When implemented, it will have to be decided whether these future types of byzantine faults will result in a tombstoning (and if not, the slash amounts will not be capped by a slashing period). 

Reviewer 2: * byznatine -> byzantine
* tranasction -> transaction

4. Validator A unjails
   
We are now in a new slashing period, however we still have to keep the door open for the previous infraction, as the evidence for Infraction 2 may still come in. As the number of slashing periods increase, it creates more complexity as we have to keep track of the highest infraction amount for every single slashing period.

> Note:  Currently, according to the `slashing` module spec, a new slashing period is created every time a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed.  See issue [#3205](https://github.com/cosmos/cosmos-sdk/issues/3205) for further details.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.

The maximum number of slashing periods is the `len(UnbondingPeriod) / len(JailPeriod)`.  The current defaults in Gaia for the `UnbondingPeriod` and `JailPeriod` are 3 weeks and 2 days, respectively.  This means there could potentially be up to 11 slashing periods concurrently being tracked per validator.  If we set the `JailPeriod >= UnbondingPeriod`, we only have to track 1 slashing period (i.e not have to track slashing periods).

> Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose setting the "jail time" for a validator who commits a consensus safety fault, to `infite` (i.e. a tombstone state).  This essentially kicks the validator out of the validator set and does not allow them to re-enter the validator set.  All of their delegators (including the operator themselves) have to either unbond or redelegate away.  The validator operator can create a new validator if they would like, with a new operator key and consensus key, but they have to "re-earn" their delegations back.  To put the validator in the tombstone state, we set `DoubleSignJailEndTime` to `time.Unix(253402300800)`, the MAX time supported by Amino.

Reviewer: ```suggestion
Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose setting the "jail time" for a validator who commits a consensus safety fault, to `infinite` (i.e. a tombstone state).  This essentially kicks the validator out of the validator set and does not allow them to re-enter the validator set.  All of their delegators (including the operator themselves) have to either unbond or redelegate away.  The validator operator can create a new validator if they would like, with a new operator key and consensus key, but they have to "re-earn" their delegations back.  To put the validator in the tombstone state, we set `DoubleSignJailEndTime` to `time.Unix(253402300800)`, the MAX time supported by Amino.
```

Reviewer: Also why is `MAX` capitalized?

> By implementing the tombstone system and getting rid of the slashing period tracking, will make the `slashing` module way simpler, especially because we can remove all of the hooks defined in the `slashing` module consumed by the `staking` module (the `slashing` module still consumes hooks defined in `staking`).

Reviewer: ```suggestion
Implementing the tombstone system and getting rid of the slashing period tracking will make the `slashing` module way simpler, especially because we can remove all of the hooks defined in the `slashing` module consumed by the `staking` module (the `slashing` module still consumes hooks defined in `staking`).
```

## Further improvements / Related proposals:

> ### Single slashing amount

Reviewer: Why is this under "further improvements" - don't we rely upon this being the case at the moment?

Author: Fixed.

> Another optimization that can be made is that if we assume that all ABCI faults for Tendermint consensus are slashed at the same level, we don't have to keep track of "max slash".  Once an ABCI fault happens once, we don't have to worry about comparing potential future ones to find the max.

Reviewer: ```suggestion
Another optimization that can be made is that if we assume that all ABCI faults for Tendermint consensus are slashed at the same level, we don't have to keep track of "max slash".  Once an ABCI fault happens, we don't have to worry about comparing potential future ones to find the max.
```

# Staking Tombstone

## Abstract

> In the current implementation of the `slashing` module, when the consensus engine informs the state machine of a validator's consensus fault, the validator is partially slashed, and put into a "jail period", a period of time in which they are not allowed to rejoin the validator set.  However, because of the nature of consensus faults and ABCI, there can be a delay between an infraction occuring, and evidence of the infraction reaching the state machine (this is one of the primary reasons for the existence of the unbonding period).

Reviewer: What does the potential evidence delay have to do with ABCI or consensus faults? It's more a consequence of unpredictable gossip layer delays and the ability of validators to selectively reveal double-signatures (e.g. to infrequently-online light clients).

I think it's worth characterizing some of those scenarios in more detail (doesn't have to be in this spec though).

One of the "design desires" of `slashing` module is that if multiple infractions occur before evidence is executed (and a validator is put in jail), they should only be punished for single worst infraction, but not cumulatively.  For example, if the sequence of events is:
1. Validator A commits Infraction 1 (worth 30% slash)
2. Validator A commits Infraction 2 (worth 40% slash)
3. Validator A commits Infraction 3 (worth 35% slash)
4. Evidence for Infraction 1 reaches state machine (and validator is put in jail)
5. Evidence for Infraction 2 reaches state machine
6. Evidence for Infraction 3 reaches state machine
   
Only Infraction 2 should have it's slash take effect, as it is the highest.  This is done, so that in the case of the compromise of a validator's consensus key, they will only be punished once, even if the hacker double-signs many blocks.  Because, the unjailing has to be done with the validator's account key, they have a chance to resecure their consensus key, and then signal that they are ready using their account key.  We call this period during which we track only the max infraction, the "slashing period".

> > Note:  Currently, from what I can tell in the code, while the stake actively delegated to a validator follows the Max slash rule, but stake that is unbonding or redelegating away from a validator gets slashed cumulatively for consecutive faults.  I think this may be a bug that needs to be fixed independently of whether the tombstone is accepted or not.

Reviewer: I don't think this is true; the slashing periods caps slashing per-validator before any redelegations or unbonding delegations are slashed.

It is true that redelegating stake can be liable for two validators' infractions at once, if that's what you mean.

(I could be wrong though - can you point to what part of the code you read?)

Author: Oh never, mind.  I realized that the `revisedFraction` returned by the `capBySlashingPeriod` function in `x/slashing` isn't the new MaxSlash amount, but rather the difference b/w MaxSlash and CurrentSlash.  So never mind, it works correctly right now.

Once, a validator rejoins by unjailing themselves, we begin a new slashing period; if they commit a new infraction after unjailing, it gets slashed cumulatively on top of the worst infraction from the previous slashing period.

However, while infractions are grouped based off of the slashing periods, because evidence can be submitted up to an `unbondingPeriod` after the infraction, we still have to allow for evidence to be submitted for previous slashing periods.  For example, if the sequence of events is:
1. Validator A commits Infraction 1 (worth 30% slash)
2. Validator A commits Infraction 2 (worth 40% slash)
3. Evidence for Infraction 1 reaches state machine (and Validator A is put in jail)
4. Validator A unjails
   
We are now in a new slashing period, however we still have to keep the door open for the previous infraction, as the evidence for Infraction 2 may still come in. As the number of slashing periods increase, it creates more complexity as we have to keep track of the highest infraction amount for every single slashing period.

> > Note:  Currently, according to the `slashing` module spec, a new slashing period is created everytime a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed, as in the current system, let's say I compromised the key of the rank 100 validator, I could bond my own validator into and out of the validator set many times, in order to create as many slashing periods I want for the validator.  Then I can create infractions for each of the slashing periods I created for the validator, allowing me to get them multiply slashed. I'm not sure if this is how it is implemented in the code, or is just a mistake in the spec.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.

Reviewer: That seems plausible; if you compromised a validator's signing key and also controlled whether that validator was in the top hundred you could abuse this mechanic (although note that you would require separate stake, you can't alter the validator's own stake bonding using only their signing key).

Definitely we should fix this.

Author: #3205

The maximum number of slashing periods is the `len(UnbondingPeriod) / len(JailPeriod)`.  The current defaults in Gaia for the `UnbondingPeriod` and `JailPeriod` are 3 weeks and 2 days, respectively.  This means there could potentially be up to 11 slashing periods concurrently being tracked per validator.  If we set the `JailPeriod >= UnbondingPeriod`, we only have to track 1 slashing period (i.e not have to track slashing periods).

> Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults, are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose that instead of being put in a "jailed state" after evidence for a consensus safety fault, validators are instead put into a "tombstone state", which means the validator is kicked out of the validator set and not allowed to rejoin.  All of the stake that was delegated to it is put into an unbonding period.  The validator operator can create a new validator if they would like, preferably with a new consensus key (do we need to enforce this?  No rational validator should reuse the same compromised key lol), but they have to "reearn" their delegations back.

Reviewer: I think we should enforce new consensus key, it's slightly cleaner to implement and definitely a good idea for the validator.

Reviewer: We also have to enforce a new operator key, in order for the redelegation to be meaningful.

> Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults, are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose that instead of being put in a "jailed state" after evidence for a consensus safety fault, validators are instead put into a "tombstone state", which means the validator is kicked out of the validator set and not allowed to rejoin.  All of the stake that was delegated to it is put into an unbonding period.  The validator operator can create a new validator if they would like, preferably with a new consensus key (do we need to enforce this?  No rational validator should reuse the same compromised key lol), but they have to "reearn" their delegations back.

Reviewer 3: `Thus, we propose that instead of being put in a "jailed state" after evidence for a consensus safety fault, validators are instead put into a "tombstone state", which means the validator is kicked out of the validator set and not allowed to rejoin.`

This seems almost identical to your definition of the jailed-state in the abstract.

Author: The key differency is _"not allowed to rejoin."_

> Doing this tombstone system and getting rid of the slashing period tracking, will make the `slashing` module way simpler, especially because we can remove the hooks between the `stake` and `slashing` modules.

Reviewer: Not entirely - we still need hooks for pubkey tracking and uptime state initialization - but we can get rid of the slashing-period-related hooks, yes.

Author: Oh I meant the hooks defined in the slashing module.  We'd still need the hooks in the `stake` module.

Reviewer: Hooks for pubkey tracking and uptime state initialization are both in the slashing module.

> > Note: The tombstone concept, only applies to byzantine faults reported over ABCI.  For slashable offenses tracked by the state machine (such as liveness faults), as there is not a delay between infraction and slashing, no slashing period tracking is needed. Also, a liveness bug probably isn't so egregious that it mandates force unbonding all delegations, and so the current jail system is adequate.

Reviewer: Agreed. I think whether we even need to slash for liveness faults is an open question - a day or two of jail (and no rewards, and social punishment) may be sufficient.

Author: Problem I have with that is that it doesn't punish the delegators.  Yes, the validator loses reputation and is kicked out of ValSet for a few days (so loses commission), but the delegators (technically including the operator too), can just redelegate and continue earning rewards.

Reviewer: That's true. Maybe we should block redelegations from jailed validators for some time after they're first jailed - just need to store the time of entering jail, not too hard.

Reviewer: Hmm maybe - can we safely calculate that cost (by extrapolating inflation over the jail period I suppose)?

## Further improvements / Related proposals:

### Single slashing amount

Another optimization that can be made is that if we assume that all ABCI faults for Tendermint consensus are slashed at the same level, we don't have to keep track of "max slash".  Once an ABCI fault happens once, we don't have to worry about comparing potential future ones to find the max.

I believe current Tendermint ABCI faults are primarily:

> - Unjustified precommits (double signs)

Reviewer: To my knowledge this is the only kind of evidence which Tendermint reports.

> - Signing a precommit when you're in unbonding phase (can be used to trick light clients)

Reviewer: Tendermint does not yet report this as evidence, which makes bisection lite client verification unsafe.

(it should report this as evidence)

Do we want to punish these two faults at different levels?  If not, we can enact the above change.  Note:  This change may make sense for current Tendermint consensus, but maybe not for a different consensus algorithm or future versions of Tendermint that may want to punish at different levels (for example, partial slashing).

### Store infractions in state instead of iterating over unbonds/redelegations

> Currently, every time evidence of a new fault comes in, we currently iterate over all of the unbonds/redelegations away from a validator to see if the slash affects them or not.  If it does, we decrease the "balance" of the `ubd` or `red`.  However, as the number of unbonds or redelegations can be very high, this might be very expensive.  Instead, we can store evidences for all infractions that happened in the last `Unbonding Period` in state, and then whenever a ubd or red hits maturity, it can check if it needs to be slashed by checking it against the last infraction that happened before they started unbonding/redelegating away.  Because we only need to store the infractions from the last unbon  

Reviewer: The problem is that - for redelegations - we need to change the amount of bonded stake of the target validator, and we need to do that at the time of the slash (otherwise they're not really "equally accountable").

Reviewer: Also it looks like the last sentence is in progress?

> <!-- 

Reviewer: I agree. Presently this is not true. See further discussion and related at:

- https://github.com/cosmos/cosmos-sdk/issues/1440
- https://github.com/cosmos/cosmos-sdk/issues/1378
- https://github.com/cosmos/cosmos-sdk/issues/1348

1. Validator A commits Infraction 1 (worth 30% slash)
2. Validator A commits Infraction 2 (worth 40% slash)
3. Evidence for Infraction 1 reaches state machine (and Validator A is put in jail)
4. Validator A unjails
   
We are now in a new slashing period, however we still have to keep the door open for the previous infraction, as the evidence for Infraction 2 may still come in. As the number of slashing periods increase, it creates more complexity as we have to keep track of the highest infraction amount for every single slashing period.

> Note:  Currently, according to the `slashing` module spec, a new slashing period is created every time a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed.  See issue [#3205](https://github.com/cosmos/cosmos-sdk/issues/3205) for further details.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.

> The maximum number of slashing periods is the `len(UnbondingPeriod) / len(JailPeriod)`.  The current defaults in Gaia for the `UnbondingPeriod` and `JailPeriod` are 3 weeks and 2 days, respectively.  This means there could potentially be up to 11 slashing periods concurrently being tracked per validator.  If we set the `JailPeriod >= UnbondingPeriod`, we only have to track 1 slashing period (i.e not have to track slashing periods).

Author: This is addressed further down, where we claim that for now, all byzantine faults have equivalent slash amounts.

Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose setting the "jail time" for a validator who commits a consensus safety fault, to `infite` (i.e. a tombstone state).  This essentially kicks the validator out of the validator set and does not allow them to re-enter the validator set.  All of their delegators (including the operator themselves) have to either unbond or redelegate away.  The validator operator can create a new validator if they would like, with a new operator key and consensus key, but they have to "re-earn" their delegations back.  To put the validator in the tombstone state, we set `DoubleSignJailEndTime` to `time.Unix(253402300800)`, the MAX time supported by Amino.

> By implementing the tombstone system and getting rid of the slashing period tracking, will make the `slashing` module way simpler, especially because we can remove all of the hooks defined in the `slashing` module.

Author: It removes the hooks in the `slashing` -> `staking` direction.

# Staking Tombstone

## Abstract

In the current implementation of the `slashing` module, when the consensus engine informs the state machine of a validator's consensus fault, the validator is partially slashed, and put into a "jail period", a period of time in which they are not allowed to rejoin the validator set.  However, because of the nature of consensus faults and ABCI, there can be a delay between an infraction occuring, and evidence of the infraction reaching the state machine (this is one of the primary reasons for the existence of the unbonding period).

In the current system design, once a validator is put in the jail for a consensus fault, after the `JailPeriod` they are allowed to send a transaction to `unjail` themselves, and thus rejoin the validator set.

> One of the "design desires" of `slashing` module is that if multiple infractions occur before evidence is executed (and a validator is put in jail), they should only be punished for single worst infraction, but not cumulatively.  For example, if the sequence of events is:

Reviewer 4: We should explicitly lay out a list of design desires for slashing and reference it here.

Author: That seems like it should be _somewhere_ in the slashing spec, but not necessarily in this tombstone spec.

> One of the "design desires" of `slashing` module is that if multiple infractions occur before evidence is executed (and a validator is put in jail), they should only be punished for single worst infraction, but not cumulatively.  For example, if the sequence of events is:

Reviewer 3: ```suggestion
One of the "design desires" of the `slashing` module is that if multiple infractions occur before evidence is executed (and a validator is put in jail), they should only be punished for single worst infraction, but not cumulatively.  For example, if the sequence of events is:
```

1. Validator A commits Infraction 1 (worth 30% slash)
2. Validator A commits Infraction 2 (worth 40% slash)
3. Validator A commits Infraction 3 (worth 35% slash)
4. Evidence for Infraction 1 reaches state machine (and validator is put in jail)
5. Evidence for Infraction 2 reaches state machine
6. Evidence for Infraction 3 reaches state machine
   

> Only Infraction 2 should have it's slash take effect, as it is the highest.  This is done, so that in the case of the compromise of a validator's consensus key, they will only be punished once, even if the hacker double-signs many blocks.  Because, the unjailing has to be done with the validator's operator key, they have a chance to re-secure their consensus key, and then signal that they are ready using their operator key.  We call this period during which we track only the max infraction, the "slashing period".

Reviewer 4: `it's` -> `its`

Once, a validator rejoins by unjailing themselves, we begin a new slashing period; if they commit a new infraction after unjailing, it gets slashed cumulatively on top of the worst infraction from the previous slashing period.

However, while infractions are grouped based off of the slashing periods, because evidence can be submitted up to an `unbondingPeriod` after the infraction, we still have to allow for evidence to be submitted for previous slashing periods.  For example, if the sequence of events is:
1. Validator A commits Infraction 1 (worth 30% slash)
2. Validator A commits Infraction 2 (worth 40% slash)
3. Evidence for Infraction 1 reaches state machine (and Validator A is put in jail)
4. Validator A unjails
   
We are now in a new slashing period, however we still have to keep the door open for the previous infraction, as the evidence for Infraction 2 may still come in. As the number of slashing periods increase, it creates more complexity as we have to keep track of the highest infraction amount for every single slashing period.

> > Note:  Currently, according to the `slashing` module spec, a new slashing period is created everytime a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed.  See issue #3205 for further details.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.

Reviewer 3: ```suggestion
> Note:  Currently, according to the `slashing` module spec, a new slashing period is created every time a validator is unbonded then rebonded.  This should probably be changed to jailed/unjailed.  See issue [#3205](https://github.com/cosmos/cosmos-sdk/issues/3205) for further details.  For the remainder of this, I will assume that we only start a new slashing period when a validator gets unjailed.
```

> The maximum number of slashing periods is the `len(UnbondingPeriod) / len(JailPeriod)`.  The current defaults in Gaia for the `UnbondingPeriod` and `JailPeriod` are 3 weeks and 2 days, respectively.  This means there could potentially be up to 11 slashing periods concurrently being tracked per validator.  If we set the `JailPeriod >= UnbondingPeriod`, we only have to track 1 slashing period (i.e not have to track slashing periods).

Reviewer 4: making the jail period >= unbonding period is unrealistic for liveness faults - I'm assuming this is only the jail period for double signing (maybe worth mentioning jail periods are distinct somewhere)

> Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults, are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose that instead of being put in a "jailed state" after evidence for a consensus safety fault, validators are instead put into a "tombstone state", which means the validator is kicked out of the validator set and not allowed to rejoin.  All of the stake that was delegated to it is put into an unbonding period.  The validator operator can create a new validator if they would like, preferably with a new consensus key (do we need to enforce this?  No rational validator should reuse the same compromised key lol), but they have to "reearn" their delegations back.

Reviewer 3: ```suggestion
Currently, in the jail period implementation, once a validator unjails, all of their delegators who are delegated to them (haven't unbonded / redelegated away), stay with them.  Given that consensus safety faults are so egregious (way more so than liveness faults), it is probably prudent to have delegators not "auto-rebond" to the validator. Thus, we propose that instead of being put in a "jailed state" after evidence for a consensus safety fault, validators are instead put into a "tombstone state", which means the validator is kicked out of the validator set and not allowed to rejoin.  All of the stake that was delegated to it is put into an unbonding period.  The validator operator can create a new validator if they would like, preferably with a new consensus key (do we need to enforce this?  No rational validator should reuse the same compromised key), but they have to "re-earn" their delegations back.
```

Given that in the current state machine, an unbonding validator cannot "rebond" and re-enter the validator set, the easiest way to accomplish the tombstone, is to just put the validator into the unbonding state.  This way, the validator cannot re-enter the validator set.  The validator operator can rejoin the validator set using a new Operator account, or wait for the unbonding to finish, and rebond with the same Operator Key.  Either way, however, their old delegations do not carry over.

> Doing this tombstone system and getting rid of the slashing period tracking, will make the `slashing` module way simpler, especially because we can remove all of the hooks defined in the `slashing` module.

Reviewer 4: "Doing this" kinda sounds funny maybe "By Implementing the"

> > Note: The tombstone concept, only applies to byzantine faults reported over ABCI.  For slashable offenses tracked by the state machine (such as liveness faults), as there is not a delay between infraction and slashing, no slashing period tracking is needed. Also, a liveness bug probably isn't so egregious that it mandates force unbonding all delegations, and so the current jail system is adequate.

Reviewer 4: This distinction between jail periods should probably be moved further up in the discussion per my previous comment

Reviewer 3: ++
